### ISSUE TITLE
notebook comms "Basic Usage.ipynb" broken

### ISSUE THREAD

**Issue by bryevdv** [bokeh/bokeh#5959]
This used to work as of `0.12.4`, does not work with `0.12.5dev14`

<img width="680" alt="screen shot 2017-03-06 at 10 29 17" src="https://cloud.githubusercontent.com/assets/1078448/23619000/d442a592-0257-11e7-9392-deb8f48b2f29.png">

@mattpap could your recent PR have affected this? I'm not aware of any other layout related PRs. 

In general, plots must continue to be multiply-"showable" in notebooks

**Comment by mattpap** [comment]
> could your recent PR have affected this?

Definitively, but that's good. I will handle this.

**Comment by mattpap** [comment]
This will be fun. You can reproduce this without notebook with:
```
from bokeh.plotting import figure, show
from bokeh.layouts import row

p1 = figure()
p1.line([1, 2, 3], [1, 2, 3])

p2 = figure()
p2.line([1, 2, 3], [1, 4, 9])

show(row(p1, p2))
show(p1)
```
What happens is that you can have plots repeated in a document, which means that layout's solver will collect variables and constraints multiple times and this results in "duplicate edit variable". Previously this error must have been conveniently hidden behind `try-catch` block. This issue is even more prominent in the notebook, because all cells share one solver.

**Comment by bryevdv** [comment]
We should probably have a new document per output cell, to be honest.

**Comment by mattpap** [comment]
That's one solution, but I'm not sure if this wouldn't affect comms in any way and perhaps could result in explosion in number of documents floating around. It would also undermine the idea of roots. 

Another idea is to have one solver per root. After all, what's a document with multiple roots, a row, a column or else? Having a solver per root layoutable would open room for having embedded layouts, e.g. in tabs, which currently are badly broken by reusing the global solver.

**Comment by bryevdv** [comment]
> Another idea is to have one solver per root. 

Actually, this has been discussed, and I thought, implemented already? But maybe I am forgetting, and it used to be a solver-per-page and now its solver-per-doc. I don't personally think more docs is big problem, but solver per-root seems worth investigating.

**Comment by bryevdv** [comment]
Thinking more, I think solver per root is definitely advised. I think having bokeh apps that can be templated ala `components` will require that. I still think there was at least a branch in the past that implemented solver per-root. @birdsarah do you remember?

Also adding `critical`, as I would not release until this is fixed.

**Comment by birdsarah** [comment]
if i'm recalling correctly, it's solver per document at the moment. document then does the add_root.

i don't recall ever implementing solver per-root.

**Comment by mattpap** [comment]
@bryevdv, I added support for per-root layout solvers in PR #5976.

**Comment by bryevdv** [comment]
Believe this was fixed by #6037
